[RPC] Allow rref.rpc_async() to be non blocking

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52167 [RPC] Move rref proxy timeout tests to general RpcTest
* **#52166 [RPC] Allow rref.rpc_async() to be non blocking**

In RRef proxy, `rref.rpc_async()` has some blocking behavior due to blocking on getting the RRef type. This can cause performance hits and it would be good to make this truly async.

This diff changes the call to `_get_type()` to return a future, and then runs the actual `rpc_async` call as a callback. We use a `torch.futures.Future()` to store the result of RPC API so that we can return it without incurring any blocking behavior. 

Differential Revision: [D25944751](https://our.internmc.facebook.com/intern/diff/D25944751/)

rohan-varma committed:
[RPC] Allow rref.rpc_async() to be non blocking

Per title

Differential Revision: [D25944751](https://our.internmc.facebook.com/intern/diff/D25944751/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[RPC] Allow rref.rpc_async() to be non blocking"


Per title

Differential Revision: [D25944751](https://our.internmc.facebook.com/intern/diff/D25944751/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[RPC] Allow rref.rpc_async() to be non blocking"


In RRef proxy, `rref.rpc_async()` has some blocking behavior due to blocking on getting the RRef type. This can cause performance hits and it would be good to make this truly async.

This diff changes the call to `_get_type()` to return a future, and then runs the actual `rpc_async` call as a callback. We use a `torch.futures.Future()` to store the result of RPC API so that we can return it without incurring any blocking behavior. 

Differential Revision: [D25944751](https://our.internmc.facebook.com/intern/diff/D25944751/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[RPC] Allow rref.rpc_async() to be non blocking"


In RRef proxy, `rref.rpc_async()` has some blocking behavior due to blocking on getting the RRef type. This can cause performance hits and it would be good to make this truly async.

This diff changes the call to `_get_type()` to return a future, and then runs the actual `rpc_async` call as a callback. We use a `torch.futures.Future()` to store the result of RPC API so that we can return it without incurring any blocking behavior. 

Differential Revision: [D25944751](https://our.internmc.facebook.com/intern/diff/D25944751/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[RPC] Allow rref.rpc_async() to be non blocking"


In RRef proxy, `rref.rpc_async()` has some blocking behavior due to blocking on getting the RRef type. This can cause performance hits and it would be good to make this truly async.

This diff changes the call to `_get_type()` to return a future, and then runs the actual `rpc_async` call as a callback. We use a `torch.futures.Future()` to store the result of RPC API so that we can return it without incurring any blocking behavior. 

Differential Revision: [D25944751](https://our.internmc.facebook.com/intern/diff/D25944751/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[RPC] Allow rref.rpc_async() to be non blocking"


In RRef proxy, `rref.rpc_async()` has some blocking behavior due to blocking on getting the RRef type. This can cause performance hits and it would be good to make this truly async.

This diff changes the call to `_get_type()` to return a future, and then runs the actual `rpc_async` call as a callback. We use a `torch.futures.Future()` to store the result of RPC API so that we can return it without incurring any blocking behavior. 

Differential Revision: [D25944751](https://our.internmc.facebook.com/intern/diff/D25944751/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[RPC] Allow rref.rpc_async() to be non blocking"


In RRef proxy, `rref.rpc_async()` has some blocking behavior due to blocking on getting the RRef type. This can cause performance hits and it would be good to make this truly async.

This diff changes the call to `_get_type()` to return a future, and then runs the actual `rpc_async` call as a callback. We use a `torch.futures.Future()` to store the result of RPC API so that we can return it without incurring any blocking behavior. 

Differential Revision: [D25944751](https://our.internmc.facebook.com/intern/diff/D25944751/)

[ghstack-poisoned]

pritamdamania87 commented on torch/distributed/rpc/rref_proxy.py:
```diff
@@ -15,24 +15,66 @@ def _local_invoke_async_execution(rref, func_name, args, kwargs):
+    # if rpc_sync, allow get_type to block, else, add invoke logic as a callback
+    block_on_type = rpc_api in [torch.distributed.rpc.rpc_sync, torch.distributed.rpc.remote]
+    rref_type = rref._get_type(timeout=timeout, blocking=block_on_type)
+    should_wait_on_rref_type = isinstance(rref_type, torch._C.Future)
```
Can we always make the call to  _get_type nonblocking? That way the code is same for all APIs (rpc_sync, rpc_async and remote). The code is easier to read that way and we don't change any guarantees we provide to the user since the rpc_api call ensures we block appropriately when needed.

rohan-varma replied:
The main use case was our internal training that uses rref.rpc_async extensively, although, I will test it with remote as well

rohan-varma replied:
Updated diff to always make it non-blocking by default.

pritamdamania87 commented on torch/distributed/rpc/rref_proxy.py:
```diff
@@ -15,24 +15,66 @@ def _local_invoke_async_execution(rref, func_name, args, kwargs):
-        timeout=timeout
-    )
+    # if rpc_sync, allow get_type to block, else, add invoke logic as a callback
+    block_on_type = rpc_api in [torch.distributed.rpc.rpc_sync, torch.distributed.rpc.remote]
```
Why does `remote` block on type as well? Shouldn't `remote` be async as well?

rohan-varma replied:
I think we have to block for remote since if we run this as a callback, we would need to return a "future" to an RRef but this API expects an RRef to be returned. 

If we have a future f and we need f.result() to create an RRef, I'm not sure we currently have a good way other than:
```
f.wait()
return RRef(f.result())
```

which is what we do here, although it's blocking

rohan-varma committed:
Update on "[RPC] Allow rref.rpc_async() to be non blocking"


In RRef proxy, `rref.rpc_async()` has some blocking behavior due to blocking on getting the RRef type. This can cause performance hits and it would be good to make this truly async.

This diff changes the call to `_get_type()` to return a future, and then runs the actual `rpc_async` call as a callback. We use a `torch.futures.Future()` to store the result of RPC API so that we can return it without incurring any blocking behavior. 

Differential Revision: [D25944751](https://our.internmc.facebook.com/intern/diff/D25944751/)

[ghstack-poisoned]

rohan-varma commented on torch/distributed/rpc/rref_proxy.py:
```diff
@@ -15,24 +15,82 @@ def _local_invoke_async_execution(rref, func_name, args, kwargs):
+    rref_type = rref._get_type(timeout=timeout, blocking=False)
+    # Controls whether we are running in a callback or inline in invoke_on_owner
+    # to appropriately handle errors.
+    block_on_type = rpc_api in [
```
Still need this for branching in the callback, since it controls whether to throw the error or wrap it in a future. Although, not sure if there's a better way

rohan-varma replied:
actually, a better name for this is probably `invoke_inline`

rohan-varma committed:
Update on "[RPC] Allow rref.rpc_async() to be non blocking"


In RRef proxy, `rref.rpc_async()` has some blocking behavior due to blocking on getting the RRef type. This can cause performance hits and it would be good to make this truly async.

This diff changes the call to `_get_type()` to return a future, and then runs the actual `rpc_async` call as a callback. We use a `torch.futures.Future()` to store the result of RPC API so that we can return it without incurring any blocking behavior. 

Differential Revision: [D25944751](https://our.internmc.facebook.com/intern/diff/D25944751/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[RPC] Allow rref.rpc_async() to be non blocking"


In RRef proxy, `rref.rpc_async()` has some blocking behavior due to blocking on getting the RRef type. This can cause performance hits and it would be good to make this truly async.

This diff changes the call to `_get_type()` to return a future, and then runs the actual `rpc_async` call as a callback. We use a `torch.futures.Future()` to store the result of RPC API so that we can return it without incurring any blocking behavior. 

Differential Revision: [D25944751](https://our.internmc.facebook.com/intern/diff/D25944751/)

[ghstack-poisoned]

rohan-varma commented:
@pritamdamania87 @mrshenli This one should be ready for another round of review whenever you get a chance

pritamdamania87 commented on torch/distributed/rpc/rref_proxy.py:
```diff
@@ -15,24 +15,82 @@ def _local_invoke_async_execution(rref, func_name, args, kwargs):
+        # Create a future whose result is the result of the above rpc_api()
+        # future. We don't simply return rref_type.then(cb) since the user would
+        # have to call wait() twice to access the future result.
+        ret_fut = torch.futures.Future()  # type: ignore
```
To make `remote` async as well, you could create an RRef in Python and then have a private RPC `remote` API that accepts a python created user rref instead of creating one internally. Then you can just return this RRef to the user and still have async behavior for `remote`.

pritamdamania87 reviewed (commented):
Sorry about the delay on this PR, I forgot to follow up on this.
